File: src/Sample/Plumix.Sample/Demos/General/RenderingDebugFlagsDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/rendering_debug_flags_demo_page.dart (exact sample parity)

public sealed class RenderingDebugFlagsDemoPage : StatefulWidget
{
    public override State CreateState() => new RenderingDebugFlagsDemoPageState();
}

internal sealed class RenderingDebugFlagsDemoPageState : State
{
    private int _repaintToken;

    public override void Dispose()
    {
        RenderingDebug.PaintSizeEnabled = false;
        RenderingDebug.PaintBaselinesEnabled = false;
        RenderingDebug.PaintPointersEnabled = false;
        RenderingDebug.PaintLayerBordersEnabled = false;
        RenderingDebug.DisableClipLayers = false;
        RenderingDebug.DisableOpacityLayers = false;
        RenderingDebug.DisablePhysicalShapeLayers = false;
        base.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("rendering/debug.dart flags", fontSize: 20, color: Colors.Black),
                new Text(
                    "Each toggle flips the matching library-level debug variable and rebuilds the probe below.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Wrap(
                    spacing: 8,
                    runSpacing: 8,
                    children:
                    [
                        BuildToggle(
                            "paint size",
                            RenderingDebug.PaintSizeEnabled,
                            () => RenderingDebug.PaintSizeEnabled = !RenderingDebug.PaintSizeEnabled),
                        BuildToggle(
                            "baselines",
                            RenderingDebug.PaintBaselinesEnabled,
                            () => RenderingDebug.PaintBaselinesEnabled = !RenderingDebug.PaintBaselinesEnabled),
                        BuildToggle(
                            "pointers",
                            RenderingDebug.PaintPointersEnabled,
                            () => RenderingDebug.PaintPointersEnabled = !RenderingDebug.PaintPointersEnabled),
                        BuildToggle(
                            "layer borders",
                            RenderingDebug.PaintLayerBordersEnabled,
                            () => RenderingDebug.PaintLayerBordersEnabled = !RenderingDebug.PaintLayerBordersEnabled),
                        BuildToggle(
                            "no clips",
                            RenderingDebug.DisableClipLayers,
                            () => RenderingDebug.DisableClipLayers = !RenderingDebug.DisableClipLayers),
                        BuildToggle(
                            "no opacity",
                            RenderingDebug.DisableOpacityLayers,
                            () => RenderingDebug.DisableOpacityLayers = !RenderingDebug.DisableOpacityLayers),
                        BuildToggle(
                            "no shadows",
                            RenderingDebug.DisablePhysicalShapeLayers,
                            () => RenderingDebug.DisablePhysicalShapeLayers =
                                !RenderingDebug.DisablePhysicalShapeLayers),
                    ]),
                new Expanded(child: BuildProbe()),
            ]);
    }

    private Widget BuildProbe()
    {
        return new KeyedSubtree(
            key: new ValueKey<int>(_repaintToken),
            child: new Container(
                color: Colors.White,
                padding: new Thickness(16),
                child: new Column(
                    crossAxisAlignment: CrossAxisAlignment.Start,
                    spacing: 16,
                    children:
                    [
                        new Text(
                            "Padding draws its construction lines, clips get a scissors marker.",
                            fontSize: 12,
                            color: Colors.DarkSlateGray),
                        new Padding(
                            insets: new Thickness(20, 12),
                            child: new ClipRRect(
                                borderRadius: BorderRadius.Circular(12),
                                child: new Container(
                                    width: 200,
                                    height: 64,
                                    color: Color.Parse("#FFB3E5FC"),
                                    alignment: Alignment.Center,
                                    child: new Text("clipped", fontSize: 16, color: Colors.Black)))),
                        new Opacity(
                            opacity: 0.45,
                            child: new PhysicalModel(
                                color: Color.Parse("#FFFFE082"),
                                elevation: 8,
                                borderRadius: BorderRadius.Circular(8),
                                child: new SizedBox(
                                    width: 200,
                                    height: 56,
                                    child: new Center(
                                        child: new Text(
                                            "elevated + 45% opacity",
                                            fontSize: 14,
                                            color: Colors.Black))))),
                        new Listener(
                            behavior: HitTestBehavior.Opaque,
                            child: new Container(
                                width: 200,
                                height: 44,
                                color: Color.Parse("#FFDCE3ED"),
                                alignment: Alignment.Center,
                                child: new Text("press and hold me", fontSize: 14, color: Colors.Black))),
                    ])));
    }

    private Widget BuildToggle(string label, bool enabled, Action toggle)
    {
        return new CounterTapButton(
            label: enabled ? $"{label}: on" : $"{label}: off",
            onTap: () => SetState(() =>
            {
                toggle();
                _repaintToken += 1;
            }),
            background: enabled ? Color.Parse("#FF9FC5E8") : Color.Parse("#FFDCE3ED"),
            foreground: Colors.Black,
            fontSize: 12,
            padding: new Thickness(10, 8));
    }
}

File: dart_sample/lib/demos/general/rendering_debug_flags_demo_page.dart
import 'package:flutter/rendering.dart';
import 'package:material_ui/material_ui.dart';

import '../../counter_widgets.dart';

class RenderingDebugFlagsDemoPage extends StatefulWidget {
  const RenderingDebugFlagsDemoPage({super.key});

  @override
  State<RenderingDebugFlagsDemoPage> createState() =>
      _RenderingDebugFlagsDemoPageState();
}

class _RenderingDebugFlagsDemoPageState
    extends State<RenderingDebugFlagsDemoPage> {
  int _repaintToken = 0;

  @override
  void dispose() {
    debugPaintSizeEnabled = false;
    debugPaintBaselinesEnabled = false;
    debugPaintPointersEnabled = false;
    debugPaintLayerBordersEnabled = false;
    debugDisableClipLayers = false;
    debugDisableOpacityLayers = false;
    debugDisablePhysicalShapeLayers = false;
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'rendering/debug.dart flags',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Each toggle flips the matching library-level debug variable and '
          'rebuilds the probe below.',
          style: TextStyle(fontSize: 14, color: Color(0xFF696969)),
        ),
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: <Widget>[
            _buildToggle(
              'paint size',
              debugPaintSizeEnabled,
              () => debugPaintSizeEnabled = !debugPaintSizeEnabled,
            ),
            _buildToggle(
              'baselines',
              debugPaintBaselinesEnabled,
              () => debugPaintBaselinesEnabled = !debugPaintBaselinesEnabled,
            ),
            _buildToggle(
              'pointers',
              debugPaintPointersEnabled,
              () => debugPaintPointersEnabled = !debugPaintPointersEnabled,
            ),
            _buildToggle(
              'layer borders',
              debugPaintLayerBordersEnabled,
              () =>
                  debugPaintLayerBordersEnabled = !debugPaintLayerBordersEnabled,
            ),
            _buildToggle(
              'no clips',
              debugDisableClipLayers,
              () => debugDisableClipLayers = !debugDisableClipLayers,
            ),
            _buildToggle(
              'no opacity',
              debugDisableOpacityLayers,
              () => debugDisableOpacityLayers = !debugDisableOpacityLayers,
            ),
            _buildToggle(
              'no shadows',
              debugDisablePhysicalShapeLayers,
              () => debugDisablePhysicalShapeLayers =
                  !debugDisablePhysicalShapeLayers,
            ),
          ],
        ),
        Expanded(child: _buildProbe()),
      ],
    );
  }

  Widget _buildProbe() {
    return KeyedSubtree(
      key: ValueKey<int>(_repaintToken),
      child: Container(
        color: Colors.white,
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          spacing: 16,
          children: <Widget>[
            const Text(
              'Padding draws its construction lines, clips get a scissors '
              'marker.',
              style: TextStyle(fontSize: 12, color: Color(0xFF2F4F4F)),
            ),
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 12),
              child: ClipRRect(
                borderRadius: BorderRadius.circular(12),
                child: Container(
                  width: 200,
                  height: 64,
                  color: const Color(0xFFB3E5FC),
                  alignment: Alignment.center,
                  child: const Text(
                    'clipped',
                    style: TextStyle(fontSize: 16, color: Colors.black),
                  ),
                ),
              ),
            ),
            Opacity(
              opacity: 0.45,
              child: PhysicalModel(
                color: const Color(0xFFFFE082),
                elevation: 8,
                borderRadius: BorderRadius.circular(8),
                child: const SizedBox(
                  width: 200,
                  height: 56,
                  child: Center(
                    child: Text(
                      'elevated + 45% opacity',
                      style: TextStyle(fontSize: 14, color: Colors.black),
                    ),
                  ),
                ),
              ),
            ),
            Listener(
              behavior: HitTestBehavior.opaque,
              child: Container(
                width: 200,
                height: 44,
                color: const Color(0xFFDCE3ED),
                alignment: Alignment.center,
                child: const Text(
                  'press and hold me',
                  style: TextStyle(fontSize: 14, color: Colors.black),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildToggle(String label, bool enabled, VoidCallback toggle) {
    return CounterTapButton(
      label: enabled ? '$label: on' : '$label: off',
      onTap: () => setState(() {
        toggle();
        _repaintToken += 1;
      }),
      background: enabled ? const Color(0xFF9FC5E8) : const Color(0xFFDCE3ED),
      foreground: Colors.black,
      fontSize: 12,
      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
    );
  }
}
